### PR TITLE
CBG-117 New database, push, pull replication stats

### PR DESCRIPTION
**Pull request by adamcfraser**
https://issues.couchbase.com/browse/CBG-164

Integration test: 
http://uberjenkins.sc.couchbase.com:8080/view/Build/job/sync-gateway-integration-master/926/

Depends on:
- [ ] https://github.com/couchbase/sg-bucket/pull/38

### CODE REVIEW

**Comment by tleyden** [comment]
@adamcfraser the build failed with:

```

godeps/src/github.com/couchbase/sync_gateway/base/dcp_feed.go:191:3: unknown field 'TimeReceived' in struct literal of type sgbucket.FeedEvent

script returned exit code 2
```

**Comment by adamcfraser** [comment]
@tleyden I'd forgotten about the sg-bucket change.  Pushed up a PR on sg-bucket (linked above), and bumped the manifest here to point to the PR commit.

**Comment by tleyden** [comment]
Looks like it failed again, this time due to:

```
018-12-10T15:24:12.933-08:00 [INF] rest.TestContinuousChangesSubscription: Reset logging
--- FAIL: TestContinuousChangesSubscription (5.72s)
    blip_api_test.go:247: 
        	Error Trace:	blip_api_test.go:247
        	Error:      	Received unexpected error:
        	            	Timed out waiting after 5s
        	Test:       	TestContinuousChangesSubscription
        	Messages:   	Timed out waiting for all changes.
2018-12-10T15:24:12.933-08:00 [INF] rest.TestBlipOneShotChangesSubscription: Setup logging: level: info - keys: HTTP, Sync, SyncMsg
2018-12-10T15:24:12.933-08:00 [INF] Opening Walrus database test_data_bucket-97e09cdf50ce3e68fc3d9a58133246ef on <walrus:>
```

I'm assuming this is safe to ignore?

**Review by tleyden — CHANGES_REQUESTED** [review]
See inline

File: db/changes.go [hunk 1]
```diff
@@ -658,7 +659,10 @@ func (db *Database) SimpleMultiChangesFeed(chans base.Set, options ChangesOption
 					break waitForChanges
 				}
 
+				db.DbStats.StatsCblReplicationPull().Add(base.StatKeyPullReplicationsCaughtUp, 1)
 				waitResponse := changeWaiter.Wait()
```
[the remark this answers is not in the document]
Comment (adamcfraser) [reply db/changes.go:663]: I think it's cleaner here, and avoids needing to pass a dbcontext down into changeWaiter.

File: db/crud.go [hunk 2]
```diff
@@ -1315,9 +1324,15 @@ func (db *Database) getChannelsAndAccess(doc *document, body Body, revID string)
 
 	if db.ChannelMapper != nil {
 		// Call the ChannelMapper:
+		startTime := time.Now()
+		db.DbStats.CblReplicationPush().Add(base.StatKeySyncFunctionCount, 1)
```
[the remark this answers is not in the document]
Comment (adamcfraser) [reply db/crud.go:1328]: ChannelMapper doesn't have any db context awareness - I think it makes sense to keep it that way.

File: db/database_stats.go [hunk 3]
```diff
@@ -162,3 +159,9 @@ func (db *DatabaseContext) UpdateCalculatedStats() {
 	}
 
 }
+
+func (db *DatabaseContext) IncrementDocReads(bytes int, delta int) {
+	if bytes > 0 {
+
```
[the remark this answers is not in the document]
Comment (adamcfraser) [reply db/database_stats.go]: Looks like it didn't get cleaned up after reads were split to REST/Blip.  Will remove.

File: rest/blip_sync.go [hunk 4]
```diff
@@ -499,6 +520,14 @@ func (bh *blipHandler) handleChanges(rq *blip.Message) error {
 	output.Write([]byte("["))
 	jsonOutput := json.NewEncoder(output)
 	nWritten := 0
+
+	// Include changes messages w/ proposeChanges stats, although CBL should only be using proposeChanges
+	startTime := time.Now()
+	bh.db.DbStats.CblReplicationPush().Add(base.StatKeyProposeChangeCount, int64(len(changeList)))
+	defer func() {
+		bh.db.DbStats.CblReplicationPush().Add(base.StatKeyProposeChangeCount, time.Since(startTime).Nanoseconds())
```
[the remark this answers is not in the document]
Comment (adamcfraser) [reply rest/blip_sync.go]: Fixed

File: rest/blip_sync.go [hunk 5]
```diff
@@ -535,6 +564,14 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
 	output := bytes.NewBuffer(make([]byte, 0, 5*len(changeList)))
 	output.Write([]byte("["))
 	nWritten := 0
+
+	// proposeChanges stats
+	startTime := time.Now()
+	bh.db.DbStats.CblReplicationPush().Add(base.StatKeyProposeChangeCount, int64(len(changeList)))
+	defer func() {
+		bh.db.DbStats.CblReplicationPush().Add(base.StatKeyProposeChangeCount, time.Since(startTime).Nanoseconds())
```
[the remark this answers is not in the document]
Comment (adamcfraser) [reply rest/blip_sync.go]: Fixed.

File: db/change_cache.go [hunk 6]
```diff
@@ -647,7 +648,7 @@ func (c *changeCache) processEntry(change *LogEntry) base.Set {
 		numPending := len(c.pendingLogs)
 		base.Infof(base.KeyCache, "  Deferring #%d (%d now waiting for #%d...#%d)",
 			sequence, numPending, c.nextSequence, c.pendingLogs[0].Sequence-1)
-		changeCacheExpvars.Get("maxPending").(*base.IntMax).SetIfMax(int64(numPending))
+		c.context.DbStats.StatsCblReplicationPull().Get(base.StatKeyMaxPending).(*base.IntMax).SetIfMax(int64(numPending))
```
[the remark this answers is not in the document]
Comment (adamcfraser) [reply db/change_cache.go]: Seems reasonable - I'll add.

**Comment by adamcfraser** [comment]
Rebased on master, re-running PR build.

**Review by tleyden — APPROVED** [review]

**Pull request merged**